hr_payroll_extended/models/hr_payslip.py

```diff
@@ -0,0 +1,130 @@
+
+        # Second, detach expenses from the pay slips
+        ExpenseObj = self.env['hr.expense']
+        expenses = ExpenseObj.search([('slip_id', 'in', self.ids)])
+        if expenses:
+            expenses.write({'slip_id': False})
+
+        # Third, detach account move lines from the pay slips
+        AccountMoveLineObj = self.env['account.move.line']
+        aml = AccountMoveLineObj.search([('slip_id', 'in', self.ids)])
+        if aml:
+            aml.write({'slip_id': False})
+
+        ret = super(HrPayslip, self).compute_sheet()
+
+        # Then, re-link the invoices, the expenses
+        # and the account move lines using the criterias
+        InvoiceLineObj = self.env['account.invoice.line']
+        for payslip in self:
+            # No contract? forget about it
+            if not payslip.contract_id:
+                continue
+
+            # No user? forget about it
+            user_id = payslip.contract_id.employee_id.user_id.id
+            if not user_id:
+                continue
+            employee_id = payslip.contract_id.employee_id.id
+
+            # Look for invoice lines
+            inv_ids = []
+            filters = [
+                ('invoice_id.user_id', '=', user_id),
+                ('product_id', '!=', False),
+                ('invoice_id.state', 'in', ['open', 'paid']),
+                ('invoice_id.type', '=', 'out_invoice'),
+            ]
+            move_ids = []
+            for invl in InvoiceLineObj.search(filters):
+                if invl.invoice_id.id not in inv_ids:
+                    inv_ids.append(invl.invoice_id.id)
+                    invl.invoice_id.write({'slip_id': payslip.id})
+                if invl.invoice_id.move_id and \
+                        invl.invoice_id.move_id.id not in move_ids:
+                    move_ids.append(str(invl.invoice_id.move_id.id))
+
+            inv_lines = InvoiceLineObj.search(filters)
+            invoices = inv_lines.mapped('invoice_id')
+            invoices.write({'slip_id': payslip.id})
+            moves = invoices.mapped('move_id')
+
+            # Look for expenses
+            filters = [
+                ('employee_id', '=', employee_id),
+                ('slip_id', '=', False),
+                ('state', '=', ['done', 'post']),
+            ]
+            expenses = ExpenseObj.search(filters)
+            if expenses:
+                expenses.write({'slip_id': payslip.id})
+
+            # Look for account move lines
+            if moves:
+                move_line_ids = self.env["account.move.line"].search([
+                    ('move_id', 'in', moves.ids),
+                    ('reconciled', '=', True),
+                    ('slip_id', '=', False)]
+                ).ids
+
+                if move_line_ids:
+                    move_line_ids = [str(id) for id in move_line_ids]
+                    q = """update account_move_line
+    set slip_id=%d where id in (%s)""" % (payslip.id, ','.join(move_line_ids))
```

Comment:
You don't have to convert the ids as a string. Python does that out of the box.

... % (payslip.id, tuple(move_line_ids))